File: android/src/main/kotlin/com/compress/all/flutter_compress/SizeMath.kt
package com.compress.all.flutter_compress

/**
 * Turns a [CompressionConfig] intent (target size / bitrate / quality) into
 * concrete encoder numbers. Single source of truth so `estimate()` and
 * `compress()` never disagree.
 */
object SizeMath {

    /** Safety margin — real muxed output tends to run a little over the raw
     * bitrate*duration math (container overhead, keyframes). */
    private const val SIZE_SAFETY = 0.95

    private const val MIN_VIDEO_BPS = 100_000

    /** Compute the target video bitrate (bits/sec) from the config. */
    fun videoBitrateBps(
        config: CompressionConfig,
        durationMs: Long,
        sourceBitrateKbps: Int,
        targetHeight: Int,
    ): Int {
        config.targetSizeMB?.let { mb ->
            val totalBits = mb.toLong() * 8L * 1024L * 1024L
            val durationSec = (durationMs / 1000.0).coerceAtLeast(0.001)
            val audioBps = if (config.removeAudio) 0
            else (config.audioBitrateKbps ?: 128) * 1000
            val audioBits = audioBps * durationSec
            val videoBits = totalBits * SIZE_SAFETY - audioBits
            return (videoBits / durationSec).toInt().coerceAtLeast(MIN_VIDEO_BPS)
        }
        config.videoBitrateKbps?.let { return it * 1000 }

        // Quality as a percentage of the SOURCE bitrate. An explicit
        // qualityPercent wins over the preset tier (more specific intent).
        val percent = (config.qualityPercent ?: presetPercent(config.quality))
            .coerceIn(1, 100)
        val srcBps = sourceBitrateKbps * 1000
        if (srcBps > 0) {
            // percent <= 100 guarantees we never re-inflate above the source.
            return (srcBps * percent / 100.0).toInt().coerceAtLeast(MIN_VIDEO_BPS)
        }
        // Source bitrate unknown: fall back to a resolution-based baseline
        // (anchored so 50% == a reasonable "medium"), scaled by percent.
        val width = targetHeight * 16 / 9
        val baselineMedium = width * targetHeight * 30 * 0.12
        return (baselineMedium * percent / 50.0).toInt().coerceAtLeast(MIN_VIDEO_BPS)
    }

    /** Preset tier -> percentage of source bitrate. */
    private fun presetPercent(quality: String): Int = when (quality) {
        "high" -> 80
        "medium" -> 50
        "low" -> 30
        "veryLow" -> 15
        else -> 50
    }

    /**
     * Compute output width/height: preserve aspect, only ever scale down to the
     * requested caps, then align per [CompressionConfig.alignment].
     */
    fun targetDimensions(srcW: Int, srcH: Int, config: CompressionConfig): Pair<Int, Int> {
        if (srcW <= 0 || srcH <= 0) return srcW to srcH
        val maxW = config.maxWidth
        val maxH = config.maxHeight
        // No cap requested → keep the source exactly. Aligning here would round
        // 1080 *up* to 1088, which both breaks the "only ever scale down" promise
        // and forces a full-frame GL rescale that starves the encoder on slower
        // devices (frames get dropped, so the output is short, not just smaller).
        if (maxW == null && maxH == null) return srcW to srcH

        var w = srcW.toDouble()
        var h = srcH.toDouble()
        if (maxW != null && w > maxW) {
            val s = maxW / w; w *= s; h *= s
        }
        if (maxH != null && h > maxH) {
            val s = maxH / h; w *= s; h *= s
        }
        return align(w.toInt(), h.toInt(), config.alignment)
    }

    private fun align(w: Int, h: Int, alignment: String): Pair<Int, Int> {
        // Encoders require even dimensions; auto16 uses /16 to avoid macroblock
        // padding artifacts. Always round *down* so alignment can never upscale.
        val m = if (alignment == "auto16") 16 else 2
        fun down(v: Int): Int = (v / m * m).coerceAtLeast(m)
        return down(w) to down(h)
    }

    /**
     * Whether the source should be handed back untouched.
     *
     * Integer arithmetic on purpose: the same expression runs in Kotlin, Swift
     * and JS (CLAUDE.md §12.2), and float division would let the three drift at
     * the boundary. `minSavingsPercent = 0` reproduces the original "only when
     * the output is larger or equal" rule exactly.
     */
    fun keepsOriginal(
        compressedBytes: Long,
        originalBytes: Long,
        keepOriginalIfLarger: Boolean,
        minSavingsPercent: Int,
    ): Boolean {
        if (!keepOriginalIfLarger || originalBytes <= 0) return false
        val ceiling = originalBytes - (originalBytes * minSavingsPercent) / 100
        return compressedBytes >= ceiling
    }
}

File: darwin/flutter_compress_pro/Sources/flutter_compress_pro/SizeMath.swift
import Foundation

// swiftlint:disable identifier_name
/// Turns a `CompressionConfig` intent into concrete encoder numbers. Mirrors
/// the Android `SizeMath` so both platforms hit the same target sizes.
///
/// The short names below (`w`, `h`, `s`, `m`, `v`) match the Kotlin and Dart
/// ports line for line — CLAUDE.md §12.2 makes this file one algorithm kept in
/// three languages, and diffing the copies against each other is the only
/// practical way to catch drift. Renaming one platform's copy would cost more
/// than the terse names do.
enum SizeMath {

  private static let sizeSafety = 0.95
  private static let minVideoBps = 100_000

  /// Target video bitrate in bits/sec.
  static func videoBitrateBps(
    config: CompressionConfig,
    durationMs: Int64,
    sourceBitrateKbps: Int,
    targetHeight: Int
  ) -> Int {
    if let mb = config.targetSizeMB {
      let totalBits = Double(mb) * 8.0 * 1024.0 * 1024.0
      let durationSec = max(Double(durationMs) / 1000.0, 0.001)
      let audioBps = config.removeAudio ? 0 : (config.audioBitrateKbps ?? 128) * 1000
      let audioBits = Double(audioBps) * durationSec
      let videoBits = totalBits * sizeSafety - audioBits
      return max(Int(videoBits / durationSec), minVideoBps)
    }
    if let kbps = config.videoBitrateKbps {
      return kbps * 1000
    }

    // Quality as a percentage of the SOURCE bitrate. An explicit qualityPercent
    // wins over the preset tier (more specific intent).
    let percent = min(max(config.qualityPercent ?? presetPercent(config.quality), 1), 100)
    let srcBps = sourceBitrateKbps * 1000
    if srcBps > 0 {
      // percent <= 100 guarantees we never re-inflate above the source.
      return max(Int(Double(srcBps) * Double(percent) / 100.0), minVideoBps)
    }
    // Source bitrate unknown: resolution-based baseline (50% == "medium").
    let width = targetHeight * 16 / 9
    let baselineMedium = Double(width * targetHeight * 30) * 0.12
    return max(Int(baselineMedium * Double(percent) / 50.0), minVideoBps)
  }

  /// Preset tier -> percentage of source bitrate.
  private static func presetPercent(_ quality: String) -> Int {
    switch quality {
    case "high": return 80
    case "low": return 30
    case "veryLow": return 15
    default: return 50
    }
  }

  /// Output width/height: preserve aspect, only downscale to caps, then align.
  static func targetDimensions(srcW: Int, srcH: Int, config: CompressionConfig) -> (Int, Int) {
    if srcW <= 0 || srcH <= 0 { return (srcW, srcH) }
    // No cap requested → keep the source exactly. Aligning here would round 1080
    // *up* to 1088, breaking the "only ever scale down" promise and forcing a
    // needless full-frame rescale.
    if config.maxWidth == nil && config.maxHeight == nil { return (srcW, srcH) }

    var w = Double(srcW)
    var h = Double(srcH)
    if let maxW = config.maxWidth, w > Double(maxW) {
      let s = Double(maxW) / w
      w *= s
      h *= s
    }
    if let maxH = config.maxHeight, h > Double(maxH) {
      let s = Double(maxH) / h
      w *= s
      h *= s
    }
    return align(Int(w), Int(h), config.alignment)
  }

  /// Whether the source should be handed back untouched.
  ///
  /// Integer arithmetic on purpose: the same expression runs in Kotlin, Swift
  /// and JS (CLAUDE.md §12.2), and float division would let the three drift at
  /// the boundary. `minSavingsPercent = 0` reproduces the original "only when
  /// the output is larger or equal" rule exactly.
  static func keepsOriginal(
    compressedBytes: Int64, originalBytes: Int64,
    keepOriginalIfLarger: Bool, minSavingsPercent: Int
  ) -> Bool {
    if !keepOriginalIfLarger || originalBytes <= 0 { return false }
    let ceiling = originalBytes - (originalBytes * Int64(minSavingsPercent)) / 100
    return compressedBytes >= ceiling
  }

  private static func align(_ w: Int, _ h: Int, _ alignment: String) -> (Int, Int) {
    // Always round *down* so alignment can never upscale.
    let m = alignment == "auto16" ? 16 : 2
    func down(_ v: Int) -> Int { max(v / m * m, m) }
    return (down(w), down(h))
  }
}
// swiftlint:enable identifier_name
